File: core/src/main/java/io/github/fourlastor/game/demo/round/Round.java
package io.github.fourlastor.game.demo.round;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.github.tommyettinger.ds.ObjectList;
import io.github.fourlastor.game.demo.round.faction.Faction;
import io.github.fourlastor.game.demo.state.GameState;
import io.github.fourlastor.game.demo.state.unit.Unit;
import io.github.fourlastor.game.ui.ActorSupport;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import javax.inject.Inject;

public class Round extends RoundState {

    private final StateRouter router;
    private List<CurrentFaction> factions = null;
    private int factionCounter = 0;

    @Inject
    public Round(StateRouter router) {
        this.router = router;
    }

    @Override
    public void enter(GameState state) {
        if (factions == null) {
            factions = Arrays.stream(Faction.values())
                    .map(faction -> new CurrentFaction(state.byFaction(faction)))
                    .collect(Collectors.toCollection(ObjectList::new));
            startFirstTurn(state);
        } else {
            advanceToNextTurn(state);
        }
    }

    @Override
    public void exit(GameState state) {
        for (Unit unit : state.units) {
            if (ActorSupport.removeListeners(unit.group.image, it -> it instanceof TurnListener)) {
                state.tileAt(unit.hex).actor.setColor(Color.WHITE);
            }
        }
    }

    private void startFirstTurn(GameState state) {
        factionCounter = 0;
        startTurn(state);
    }

    private void advanceToNextTurn(GameState state) {
        CurrentFaction currentFaction = factions.get(factionCounter);
        if (currentFaction.allUnitsActed()) {
            factionCounter += 1;
        }
        if (factionCounter >= factions.size()) {
            router.round();
            return;
        }
        startTurn(state);
    }

    private void startTurn(GameState state) {
        CurrentFaction currentFaction = factions.get(factionCounter);
        currentFaction.units.stream().filter(unitInTurn -> !unitInTurn.hasActed).forEach(unitInTurn -> {
            state.tileAt(unitInTurn.unit.hex).actor.setColor(Color.PINK);
            unitInTurn.unit.group.image.addListener(new TurnListener(unitInTurn));
        });
    }

    private static class CurrentFaction {
        final List<UnitInRound> units;

        private CurrentFaction(List<Unit> units) {
            this.units = units.stream().map(UnitInRound::new).collect(Collectors.toList());
        }

        boolean allUnitsActed() {
            return units.stream().allMatch(it -> it.hasActed);
        }
    }

    private class TurnListener extends ClickListener {

        private final UnitInRound unit;

        public TurnListener(UnitInRound unit) {
            this.unit = unit;
        }

        @Override
        public void clicked(InputEvent event, float x, float y) {
            router.turn(unit);
        }
    }
}

File: core/src/main/java/io/github/fourlastor/game/demo/round/Round.kt
package io.github.fourlastor.game.demo.round

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.Action
import com.badlogic.gdx.scenes.scene2d.EventListener
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.actions.Actions
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import io.github.fourlastor.game.demo.round.faction.Faction
import io.github.fourlastor.game.demo.state.GameState
import io.github.fourlastor.game.demo.state.unit.Mon
import io.github.fourlastor.game.ui.ActorSupport.removeListeners
import java.util.function.Consumer
import javax.inject.Inject

class Round @Inject constructor(private val router: StateRouter, private val stage: Stage) : RoundState() {
    private var factions: List<CurrentFaction>? = null
    private var factionCounter = 0
    override fun enter(state: GameState) {
        val currentFactions = factions
        if (currentFactions == null) {
            stage.addAction(startOfRound(state))
        } else {
            advanceToNextTurn(state, currentFactions)
        }
    }

    private fun startOfRound(state: GameState): Action {
        val currentFactions = Faction.values().map { CurrentFaction(state.byFaction(it)) }
        factions = Faction.values().map { CurrentFaction(state.byFaction(it)) }
        val actions = currentFactions
            .flatMap { it.units.map { unitInRound -> unitInRound.mon.onRoundStart() } }
        return Actions.sequence(*actions.toTypedArray(), Actions.run { startFirstTurn(state, currentFactions) })
    }

    override fun exit(state: GameState) {
        for (tile in state.tiles) {
            if (removeListeners(tile.actor) { it: EventListener? -> it is TurnListener }) {
                tile.actor.color = Color.WHITE
            }
        }
    }

    private fun startFirstTurn(state: GameState, currentFactions: List<CurrentFaction>) {
        factionCounter = 0
        startTurn(state, currentFactions)
    }

    private fun advanceToNextTurn(state: GameState, currentFactions: List<CurrentFaction>) {
        val currentFaction = currentFactions[factionCounter]
        if (currentFaction.allUnitsActed()) {
            factionCounter += 1
        }
        if (factionCounter >= currentFactions.size) {
            currentFactions.forEach(
                Consumer { faction: CurrentFaction ->
                    faction.units.forEach(
                        Consumer { unitInRound: UnitInRound -> unitInRound.mon.onRoundEnd() }
                    )
                }
            )
            router.round()
            return
        }
        startTurn(state, currentFactions)
    }

    private fun startTurn(state: GameState, currentFactions: List<CurrentFaction>) {
        val currentFaction = currentFactions[factionCounter]
        currentFaction.units.asSequence()
            .filter { unitInTurn: UnitInRound -> !unitInTurn.hasActed }
            .forEach { unitInTurn: UnitInRound ->
                val unitTile = state.tileAt(unitInTurn.mon.hex)
                unitTile.actor.color = Color.PINK
                unitTile.actor.addListener(TurnListener(unitInTurn))
            }
    }

    private class CurrentFaction(mons: List<Mon>) {
        val units: List<UnitInRound>

        init {
            this.units = mons.map { UnitInRound(it) }
        }

        fun allUnitsActed(): Boolean = units.all { it.hasActed }
    }

    private inner class TurnListener(private val unit: UnitInRound) : ClickListener() {
        override fun clicked(event: InputEvent, x: Float, y: Float) {
            router.turn(unit)
        }
    }
}
